File: csharp/rules/domain-modeling/no-unwrapping-domain-primitives.cs
namespace Thruput.Semgrep.Tests
{
    public class UnwrappingTests
    {
        public bool CompareValues(CustomerId a, CustomerId b)
        {
            // ruleid: csharp-no-unwrapping-domain-primitives
            return a.Value == b.Value;
        }

        public bool CompareIds(Order a, Order b)
        {
            // ruleid: csharp-no-unwrapping-domain-primitives
            return a.Id == b.Id;
        }

        public bool CompareDirectly(CustomerId a, CustomerId b)
        {
            // ok: csharp-no-unwrapping-domain-primitives
            return a == b;
        }

        public bool CompareEquals(Order a, Order b)
        {
            // ok: csharp-no-unwrapping-domain-primitives
            return a.Equals(b);
        }
    }

    public record CustomerId(string Value);
    public record Order(int Id);
}

File: python/rules/domain-modeling/no-unwrapping-domain-primitives.py
class CustomerId:
    def __init__(self, value: str):
        self.value = value

    def __eq__(self, other: object) -> bool:
        return isinstance(other, CustomerId) and self.value == other.value

def check_same_customer(a: CustomerId, b: CustomerId) -> bool:
    # ruleid: python-no-unwrapping-domain-primitives
    return a.value == b.value

def check_different_customer(a: CustomerId, b: CustomerId) -> bool:
    # ruleid: python-no-unwrapping-domain-primitives
    return a.value != b.value

def check_safe_customer(a: CustomerId, b: CustomerId) -> bool:
    # ok: python-no-unwrapping-domain-primitives
    return a == b
